lib/ferrum/browser/subscriber.rb

```diff
@@ -33,6 +33,24 @@ def call(message)
           block.call(params, index, total)
         end
       end
+
+      private
+
+      def handle_any_times(event, &block)
+        @on[event] << block
+      end
+
+      def handle_once(event, &block)
```

Comment:
This is very unreadable and hard, if I were you I'd start refactoring it first before moving further. Not sure if I like the idea of having this `:handler_type` on event listener. Can you explain why we need it in the first place?

Reply:
> if I were you I'd start refactoring it first before moving further

yeah, of course, it's a Draft PR with [WIP] code, just the attempt to make it works.

> Can you explain why we need it in the first place?

sure, the original intention is to follow the existing implementation in [puppeteer link that you provided](https://github.com/rubycdp/ferrum/issues/171#issue-823832394).

Those guys using `once` type of handler just for a few cases, for example:
![image](https://user-images.githubusercontent.com/23525618/141420520-a0b7f352-363b-4df1-96f7-8819fd882a24.png)
At the same time they use a handler for any_times for such more cases:
![image](https://user-images.githubusercontent.com/23525618/141420677-845ac380-f153-402b-928f-03628ea40e80.png)

Hence, they have an intention to use a handler with a `once` type for `Tracing.tracingComplete` event.

Apart from that, the `microsoft/playwright` implementation of tracing also use a handler with `once` type:
```ts
new Promise(f => this._tracingClient!.once('Tracing.tracingComplete', f)),
```
Source: https://github.com/microsoft/playwright/blob/c373986ca0b91b76d4910ce0617b55e50934114c/packages/playwright-core/src/server/chromium/crBrowser.ts#L268

I'm not sure about the mandatory of use the once-type event handler for `Tracing.tracingComplete` event, but it looks reasonable to handle events only once by `stop` command.

Reply:
I still don't understand the intention. If I subscribe to `Tracing.tracingComplete` does it mean it is executed multiple times? If so what are the implications of this?

Reply:
> what are the implications of this

great question! well, as I understand, the issue is playable for the case when we use tracing a few times for the same page instance.

For example, we wanna trace events -> then stop tracing ->, then do whatever something ->, and then start new tracing but for a different file:
```ruby
context "handle tracing.Complete event once" do
  it "outputs a trace" do
    file_path = "#{PROJECT_ROOT}/spec/tmp/trace.json"
    browser.tracing.start(path: file_path)
    browser.go_to("https://www.google.com")
    browser.tracing.stop
    sleep 1
    expect(File.exist?(file_path)).to be(true)
	
    file_path2 = "#{PROJECT_ROOT}/spec/tmp/trace2.json"
    browser.tracing.start(path: file_path2)
    browser.go_to("https://www.facebook.com")
    browser.tracing.stop
    sleep 1
    expect(File.exist?(file_path2)).to be(true)

    file_path3 = "#{PROJECT_ROOT}/spec/tmp/trace3.json"
    browser.tracing.start(path: file_path3)
    browser.go_to("https://github.com")
    browser.tracing.stop
    sleep 1
    expect(File.exist?(file_path3)).to be(true)
  ensure
    FileUtils.rm_f(file_path)
    FileUtils.rm_f(file_path2)
    FileUtils.rm_f(file_path3)
  end
end
```

here is a difference:
```ruby
def stop
  client.on("Tracing.tracingComplete") do |event|
    puts "Call stream_to_file for #{path}"
    stream_to_file(event.fetch("stream"), path: path)
  end
  client.command("Tracing.end")
end
```
```
Ferrum::Browser
  tracing
    handle tracing.Complete event once
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace.json
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace2.json
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace2.json
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
      outputs a trace
```
```ruby
def stop
  client.on("Tracing.tracingComplete", handler_type: :once) do |event|
    puts "Call stream_to_file for #{path}"
    stream_to_file(event.fetch("stream"), path: path)
  end
  client.command("Tracing.end")
end
```
```
Ferrum::Browser
  tracing
    handle tracing.Complete event once
call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace.json
call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace2.json
call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
```

Reply:
I think it makes more sense to subscribe to the events on start. Does tracingComplete have any param that it doesn’t have any more data to send?

Reply:
> Does tracingComplete have any param that it doesn’t have any more data to send?

nope:
```ruby
Ferrum::Browser
  tracing
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace.json
{"dataLossOccurred"=>false, "stream"=>"1", "traceFormat"=>"json", "streamCompression"=>"none"}
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace2.json
{"dataLossOccurred"=>false, "stream"=>"2", "traceFormat"=>"json", "streamCompression"=>"none"}
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace2.json
{"dataLossOccurred"=>false, "stream"=>"2", "traceFormat"=>"json", "streamCompression"=>"none"}
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
{"dataLossOccurred"=>false, "stream"=>"3", "traceFormat"=>"json", "streamCompression"=>"none"}
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
{"dataLossOccurred"=>false, "stream"=>"3", "traceFormat"=>"json", "streamCompression"=>"none"}
Call stream_to_file for /home/mifrill/Documents/source/rubycdp/ferrum/spec/tmp/trace3.json
{"dataLossOccurred"=>false, "stream"=>"3", "traceFormat"=>"json", "streamCompression"=>"none"}
```
But we can use use `index` to play with `once` type events! :smiley: I didn't catch it, thank you

> I think it makes more sense to subscribe to the events on start.

yeah, it makes sense